### PR TITLE
[BLIP-5] Baseline Battleship

### PR DESCRIPTION
**Pull request by skosito**
# Description

Please check BLIP-5 https://github.com/eea-oasis/baseline-blips/issues/7 and grant application https://github.com/eea-oasis/baseline-grants/issues/51 for more details.

Purpose of this work in general is to have 'hello world' type application, which should help new developers getting onboard quickly and having something tangible they can check and improve/modify. Battleship game was chosen to serve the purpose of this.

This PR contains working battleship game, between 2 players, each one placing 1 ship which opponent is trying to hit. It uses Kafka for messaging, Circom for circuit, ganache-cli, and simple in-memory registry services for organizations and workgroups. Even though organizations and workgroups were not absolutely necessary, nomenclature was kept in order to be aligned with other examples around the codebase.

Please keep in mind that this is first iteration of the game, and there are improvements to be made. There are 2 reasons for this, first is nature of grant process in this case, which took limited amount of time, and also as motivation for new people to contribute easily. We have and maintain project board here https://github.com/eea-oasis/baseline/projects/1 where backlog items contains already identified improvements that can be made in future.

Of course, since this is baseline hello world app, we will fix anything that is potentially wrong from protocol point of view as part of this grant work, and we are looking forward to feedback in this PR.

## Related Issue
BLIP issue:
https://github.com/eea-oasis/baseline-blips/issues/7 
Grant issue:
https://github.com/eea-oasis/baseline-grants/issues/51

## How Has This Been Tested

To test this you can navigate to ops folder and execute run.sh script. This will start 2 front end containers on ports 8080 and 8081, 2 baseline protocol implementations node services for each of those, ganache-cli and kafka containers, all ready and connected for playing the game.

## Types of changes

There is no proper type for this change. It is neither bugfix or feature or breaking change, but an isolated example. Selecting new feature as it is closest to what we have here, but probably should be new category here for examples.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

We did not add automated tests for this, it is future improvement. This is isolated example and it did not break anything in the codebase.
Documentation will be changed to highlight this example.

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by humbitious — DISMISSED** [review]

**Review by ognjenkurtic — APPROVED** [review]

File: examples/battleship/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+|Organization | Organization |
+|Workgroup  | Workgroup |
+|Gameplay  | Workflow |
+|Turn  | Workstep   |
+|Board | Document |
+|Ship | Record/Private Input |
+|Hit | Verified Proof |
+
+## Gameplay
+Players generate an organization and join a workgroup. Ships are privately placed on a coordinate grid on which each players ship location is marked. The position of the ship in relation to the coordinate grid is then used to generate current state. Each turn, players attempt to locate their opponents ship through a series of coordinate defined missle launches. Target coordinates and board state are used as inputs in a zero-knowledge proof to determine if the opponents ship is present in the target space. A hit will result in an updated board state and verified proof that is recorded as a leaf in a merkle tree via a transaction to a locally run blockchain. Misses simply progress gameplay. Once a players ship is sunk, the game ends and a final proof is recorded. Because each proof generation triggered by a hit contains board state and previous hits, the final game proof can be used to verify the validity of the entire game.
```
Comment (fosgate29) [inline examples/battleship/README.md:20]: a typo: missle

**Review by fosgate29 — APPROVED** [review]

**Comment by Therecanbeonlyone1969** [comment]
@skosito ... will approve PR but my comments from the excel plugin PR stand:

- smaller PRs -> super hard to review large PRs -> prefer 10 small PRs or reviewable draft PRs that take me 10 - 20 min each rather than 1 big one that takes me a day+ to do a good job
- tests!!!! in CI/CD pipeline
- externalize stuff to avoid code duplication and tight coupling

Let's discuss these fundamental things during our next core-dev meeting.

cc @humbitious -> grant opportunities to build the infra for all this; CI/CD hooks, npm packages etc.

**Review by Therecanbeonlyone1969 — DISMISSED** [review]

**Comment by skosito** [comment]
> @skosito ... will approve PR but my comments from the excel plugin PR stand:
> 
> * smaller PRs -> super hard to review large PRs -> prefer 10 small PRs or reviewable draft PRs that take me 10 - 20 min each rather than 1 big one that takes me a day+ to do a good job
> * tests!!!! in CI/CD pipeline
> * externalize stuff to avoid code duplication and tight coupling
> 
> Let's discuss these fundamental things during our next core-dev meeting.
> 
> cc @humbitious -> grant opportunities to build the infra for all this; CI/CD hooks, npm packages etc.

Hey @Therecanbeonlyone1969  thanks for review, appreciate it.
I understand your concerns and they are perfectly valid. Unfortunately, since there was a tight deadline on this project, and capacity from planning was not met during actual development, some of the things just were not possible to finish in time. So, apart from code quality that should be essential, the overall process should be considered in that discussion, organization of the project, grant process, planning, execution etc.

**Comment by ognjenkurtic** [comment]
> > @skosito ... will approve PR but my comments from the excel plugin PR stand:
> > 
> > * smaller PRs -> super hard to review large PRs -> prefer 10 small PRs or reviewable draft PRs that take me 10 - 20 min each rather than 1 big one that takes me a day+ to do a good job
> > * tests!!!! in CI/CD pipeline
> > * externalize stuff to avoid code duplication and tight coupling
> > 
> > Let's discuss these fundamental things during our next core-dev meeting.
> > cc @humbitious -> grant opportunities to build the infra for all this; CI/CD hooks, npm packages etc.
> 
> Hey @Therecanbeonlyone1969 thanks for review, appreciate it. I understand your concerns and they are perfectly valid. Unfortunately, since there was a tight deadline on this project, and capacity from planning was not met during actual development, some of the things just were not possible to finish in time. So, apart from code quality that should be essential, the overall process should be considered in that discussion, organization of the project, grant process, planning, execution etc.

Good exchange. As Andreas mentioned, let's prioritize these topics in the next core-dev meeting - we need to find a way to move forward with much smaller tasks. This will help us move faster with higher quality but also lower the barrier to entry for newcomers.

**Review by ognjenkurtic — APPROVED** [review]

**Review by humbitious — APPROVED** [review]

**Comment by skarred14** [comment]
Same comments as on BLIP for the excel addin. Smaller PRs and test driven approvals for merges going forward.

**Review by skarred14 — APPROVED** [review]

**Review by Therecanbeonlyone1969 — APPROVED** [review]
good to go for now

**Pull request merged**